cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java

```diff
@@ -93,30 +95,43 @@ private void setConfig(EntityId entityId,
       ProfileDataset profileDataset = ProfileDataset.get(context, datasetFramework);
       PreferencesDataset preferencesDataset = PreferencesDataset.get(context, datasetFramework);
 
-      // need to get old property and check if it contains profile information
-      Map<String, String> oldProperties = preferencesDataset.getPreferences(entityId);
-      preferencesDataset.setPreferences(entityId, propertyMap);
-      NamespaceId namespaceId =
-        entityId instanceof NamespacedEntityId ? ((NamespacedEntityId) entityId).getNamespaceId() : NamespaceId.SYSTEM;
+      boolean isInstanceLevel = entityId.getEntityType().equals(EntityType.INSTANCE);
+      NamespaceId namespaceId = isInstanceLevel ?
+        NamespaceId.SYSTEM : ((NamespacedEntityId) entityId).getNamespaceId();
 
       // validate the profile and publish the necessary metadata change if the profile exists in the property
       Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(namespaceId, propertyMap);
-      // get the old profile information from the previous properties
-      Optional<ProfileId> oldProfile = SystemArguments.getProfileIdFromArgs(namespaceId, oldProperties);
       if (profile.isPresent()) {
         ProfileId profileId = profile.get();
+        // if it is instance level set, the profile has to be in SYSTEM scope, so throw BadRequestException when
+        // setting a USER scoped profile
+        if (isInstanceLevel && !propertyMap.get(SystemArguments.PROFILE_NAME).startsWith("SYSTEM")) {
```

Comment:
instead of doing a prefix check, it's better to do `profileId.getNamespaceId() != NamespaceId.SYSTEM`.

Reply:
The namespace id will always be `NamespaceId.SYSTEM`, since the method in `SystemArguments` need to take a namespace id otherwise it is not able to tell from the property which namespace this profile is in. If it is instance level, we always pass `NamespaceId.SYSTEM` for it

Reply:
I see. Let's use EntityScope.SYSTEM.name() instead of "SYSTEM" then